services/snapshot/service.go

```diff
@@ -43,44 +43,64 @@ func init() {
 var empty = &protoempty.Empty{}
 
 type service struct {
-	snapshotter snapshot.Snapshotter
-	emitter     events.Poster
+	snapshotters           map[string]snapshot.Snapshotter
+	defaultSnapshotterName string
+	emitter                events.Poster
 }
 
 func newService(ic *plugin.InitContext) (interface{}, error) {
 	evts := events.GetPoster(ic.Context)
-	snapshotters, err := ic.GetAll(plugin.SnapshotPlugin)
+	rawSnapshotters, err := ic.GetAll(plugin.SnapshotPlugin)
 	if err != nil {
 		return nil, err
 	}
-	cfg := ic.Config.(*config)
-
-	sn, ok := snapshotters[cfg.Default]
-	if !ok {
-		return nil, errors.Errorf("default snapshotter not loaded: %s", cfg.Default)
-	}
-
 	md, err := ic.Get(plugin.MetadataPlugin)
 	if err != nil {
 		return nil, err
 	}
+	snapshotters := make(map[string]snapshot.Snapshotter)
+	for name, sn := range rawSnapshotters {
+		snapshotters[name] = metadata.NewSnapshotter(md.(*bolt.DB), name, sn.(snapshot.Snapshotter))
+	}
+
+	cfg := ic.Config.(*config)
+	_, ok := snapshotters[cfg.Default]
+	if !ok {
+		return nil, errors.Errorf("default snapshotter not loaded: %s", cfg.Default)
+	}
 
 	return &service{
-		snapshotter: metadata.NewSnapshotter(md.(*bolt.DB), cfg.Default, sn.(snapshot.Snapshotter)),
-		emitter:     evts,
+		snapshotters:           snapshotters,
+		defaultSnapshotterName: cfg.Default,
+		emitter:                evts,
 	}, nil
 }
 
+func (s *service) getSnapshotter(name string) (snapshot.Snapshotter, error) {
+	if name == "" {
+		name = s.defaultSnapshotterName
+	}
+	sn, ok := s.snapshotters[name]
+	if !ok {
+		return nil, errors.Errorf("snapshotter not loaded: %s", name)
```

Comment:
I am wondering if we should use one of the `errdefs` here. This will get defaulted as an unknown error back to the client but I am thinking we wanting to specify something more useful. `ErrNotFound` is probably not correct here, also `ErrUnavailable` should not be used because this error is not transient. I am wondering if this would fall into the category of `ErrFailedPrecondition` since it indicates the server is not in a state required to complete the action, and further action is required to get the server into state. Unfortunately there is nothing the client could do to fix the state, but I still think it is most relevant.
```
    //  (c) Use FailedPrecondition if the client should not retry until
    //      the system state has been explicitly fixed.  E.g., if an "rmdir"
    //      fails because the directory is non-empty, FailedPrecondition
    //      should be returned since the client should not retry unless
    //      they have first fixed up the directory by deleting files from it.
```

See https://godoc.org/google.golang.org/grpc/codes for the definitions we are trying to stick to.